[REVIEW] CSV reader input stage optimizations

OlivierNV opened the pull request:
- Fully offload row offset gathering to GPU
- Always perform IO in chunks (avoids duplicate reads of the data)
- Pave the way for escapechar support and improved pandas compatibility
- Eliminate character position host+device allocations (often bigger than the csv itself)

High-level overview of the data-parallel row offset gathering kernel: since the number of rows in a given character block may depend on the initial parser state (like whether the block starts in a middle of a quote or not), a separate row count and output parser state is computed for every possible input parser state per 16KB character block.
 The result is then used to infer the parser state and starting row at the beginning of every character block.
 A second pass can then output the location of every row (which is needed for the subsequent parallel conversion of every row from csv text to cudf binary form)

~50x speedup of row offset gathering stage, which should result in ~3x overall speedup of read_csv (content dependent, less with `quoting=QUOTE_NONE` and/or few long rows, more with small string columns, many short rows and/or large values of `skiprows`)

[Note to reviewers: I'd suggest reviewing the headers first, before the .cu/.cpp, eg: in the exact opposite than github's default alphabetical order]

Resolves #1229 , #1228

GPUtester commented:
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

harrism commented on cpp/src/io/csv/csv_gpu.h:
```diff
@@ -23,6 +23,99 @@ namespace cudf {
 namespace io {
 namespace csv {
 namespace gpu {
+
```
This should probably be a .cuh file since it includes `__device__` code.

OlivierNV replied:
> This should probably be a .cuh file since it includes `__device__` code.

It's exposing host inline functions, but they also happen to be also be used on the device side (__host__ __device__). It's not actually device-specific (everything in this file is meant to be usable on the cpp side) ~ not sure if there is a way other than `__host__ __device__` declaration to make normal code run on either cpp or cu ?

OlivierNV replied:
Actually, I might just move these, along with the other __device__ functions that are not CSV-specific to `parsing_utils.cuh`, since they will likely be needed for the equivalent optimization in the JSON parser (which would allow us to get rid of the `{find|count}_all_from_set` and related utils

OlivierNV replied:
Meh, looks like it's going to be a mess due to some things being under the `experimental` namespaces and others not (`cudf::io::gpu`, `cudf::io::csv::gpu`, `cudf::experimental::io::gpu` etc) -> will wait until the `experimental` namespace is removed when tackling the JSON reader.

vuule commented on cpp/src/io/csv/csv_gpu.h:
```diff
@@ -23,6 +23,99 @@ namespace cudf {
+ * EOF: End state (EOF reached)
+ **/
+enum { ROW_CTX_NONE = 0, ROW_CTX_QUOTE = 1, ROW_CTX_COMMENT = 2, ROW_CTX_EOF = 3 };
+
```
I'm finding it very hard to understand the gather implementation. Can you please include the description of the context format(s) in this header. As it is, it is very disjointed and hard to understand the hard-coded constants used in the implementation.

OlivierNV replied:
Added a better description of the row context. It's basically just a 2-bit parsing state with row count (current row count and current parsing state), and the packed format just maps 4 possible input states to 4 possible output states (so the parsing is done for all possible input parsing states so it can be done in parallel without requiring knowledge of the initial input state, which is the part that causes serialization).

OlivierNV replied:
[I added the description of each format above the corresponding `make_row_context` and `pack_row_contexts` helpers]

vuule commented on cpp/src/io/csv/csv_gpu.cu:
```diff
@@ -570,6 +571,315 @@ __global__ void convertCsvToGdf(const char *raw_csv,
+ * @brief merge two packed row contexts (each corresponding to a block of characters)
+ * and returns the packed row context corresponding to the merged character block
+ **/
+inline __device__ uint64_t merge_row_contexts(uint64_t first_ctx, uint64_t second_ctx)
```
AFAICT, CUDA supports bit fields. IMO the code would be _much_ more readable if we used this instead of plain uint32/64.

OlivierNV replied:
It also would be horribly inefficient

vuule replied:
I don't see why it would have to be. Is this a known issue with CUDA and bit-fields?

OlivierNV replied:
The struct would first have to be an indexable array of arbitrary bit-length values. I'd encourage you to experiment over at Godbolt.org rather than rely on assumptions.

vuule replied:
> The struct would first have to be an indexable array of arbitrary bit-length values.

I don't think I understand this. The struct could just have separate fields.

I assume that the language features are usable, it's true that sometimes it's not the case :)
I wrote up some sample code that's fairly similar to the row contexts: https://godbolt.org/z/UfwjCs
The generated code looks similar between bit-fields and manual packing. The hand-crafted version does seem slightly smaller.

OlivierNV replied:
Heh. You don't want to use a signed type for bitfields :) Half the instructions generated will be to sign-extend & mask stuff.
Regarding the indexable array is because the 2-bit context id is indexed by a variable in many places (corresponds to a shift by a variable), so one operation for example would be:
`ctxid= ctxid_out[ctxid_in]` where ctxid_out is an array of 2-bit values: I'm not sure it's even possible to declare such a thing in any high-level SW language (not a problem for HW languages), but here it's just `ctxid = (packed_ctx >> (8 + ctxid_in * 2)) & 3;` You could avoid the constant `8` in a struct, but not the shift afaik.
In most cases, the merge operations here will result in very few instructions (which is critical for perf here).

vuule replied:
> You don't want to use a signed type for bitfields :)
Yup, switching to `unsigned char` shaves off two instructions from `f1()`. Focused on field sizes and forgot to about the sign. But, indexing issues aside, this sample suggests that the bit-fields are as fast as manual packing/unpacking

I haven't dug into the code enough to see the "indexing" used. Shame, this would remove so much noise in the code.

OlivierNV replied:
Indexing is used in the simplest `get_row_context()` (the short-context version is inlined inside `merge_short_context` because it's only needed in one place)
The difference here is like an order-of-magnitude. With structs of bitfields, the best I could do on godbolt for the simplest `get_row_context` is a switch (and the merge calls this function 3 times per merge), which resulted in literally 5x more instructions and 3 data-dependent branches (big no-no for both cpu and gpu).
It also resulted in a *lot* more source code fwiw.

vuule replied:
Interesting. Do you have this saved?

OlivierNV replied:
I was curious, in case compilers got a lot better since the last time I checked: [Godbolt says "Nope"](https://godbolt.org/z/kfJG95)
Note that things tend to get worse on device code, since GPU doesn't have a lot of fancy x86 zero or sign-extension on load or partial register access -> you generally don't want to pass chars around everywhere like we do (on the other hand, ptxas does some very surprisingly clever stuff with the LOP3 instruction for series of logical operations - I still have to wrap my head around that instruction)

vuule commented on cpp/src/io/csv/csv_gpu.cu:
```diff
@@ -570,6 +571,315 @@ __global__ void convertCsvToGdf(const char *raw_csv,
+                                                  uint32_t c1  = 0,
+                                                  uint32_t c2  = 0)
+{
+  return (id0 << 8) | (id1 << 10) | (id2 << 12) | (ROW_CTX_EOF << 14) | (c0) | (c1 << 1) |
```
There's a lot of hard-coded constants. It would be great if they were named.
IMO it would be best to declare field sizes as constexpr ints, and use these constants in context struct declaration(s), and also in the implementation.
Anything derived from these sizes (masks, offsets) can be also computed at compile time so the final code will effectively be the same as the hard-coded version.

OlivierNV replied:
IMO I actually find this more readable (along with the comment immediately above), rather than defining something like:
```
enum { BIT_POS_ID0 = 8 + 0*2, BIT_POS_ID1 = 8 + 1*2, BIT_POS_ID2 = 8 + 2*2, BIT_POS_ID3 = 8 + 3 * 2 };
return (id0 << BIT_POS_ID0) | (id1 << BIT_POS_ID1) etc..
```
It really doesn't matter what the values of the constants are as long as it matches between `make_short_row_context` and `merge_short_row_context` (the only two places where these constants are used), so in that sense, I don't think it detracts from understanding the merging operation.

OlivierNV replied:
I think the comment two lines above makes it clearer than a bunch of constexpr values defined further away:
```
 * @brief Short row context:
 * 1-bit count (0 or 1) per context in the lower 4 bits
 * 2-bit output context id per input context in bits 8..15
```

OlivierNV replied:
I'll focus on adding more comments instead.

OlivierNV commented:
rerun tests

OlivierNV commented on cpp/src/io/csv/reader_impl.cu:
```diff
@@ -238,28 +232,41 @@ table_with_metadata reader::impl::read(size_t range_offset,
+    size_t header_end =
+      parse_csv_header(h_uncomp_data + header_offset, h_uncomp_size - header_offset);
+
+    if (range_offset == 0) { data_start_offset = header_end; }
```
You still need to call `parse_csv_header(h_uncomp_data + range_offset, ...)` when `range_offset != 0`.
As the comment says, it's a bit of a ugly hack that we parse the first data row as if it were the header (primarily needed to infer the number of columns). I tried to organize the code such that when we fix the datasource to allow reading the header outside the mapped range, we'll be able to just always parse the header at header_offset=0, and remove the various range_offset comparisons (rather than having two separate paths calling `parse_csv_header`).

OlivierNV replied:
Oh, I see what you mean: `parse_csv_header` does indeed has a big side effect: it updates the `header` class member (something that was previously done in `select_rows`). Not ideal, but I kept this as is in an effort to limit the code changes (the `header` class member could probably just be a local in this function).

trevorsm7 replied:
Ah, I didn't notice that `parse_csv_header` and this scope are both methods of `reader::impl`

Hm, how about something like:
```cpp
size_t const header_offset = (range_offset != 0) ? find_first_row_start(h_uncomp_data, h_uncomp_size) : 0;
size_t const header_end = parse_csv_header(h_uncomp_data + header_offset, h_uncomp_size - header_offset);
size_t const data_start_offset = (range_offset != 0) ? header_offset : header_end;
```
or
```cpp
size_t const data_start_offset = [=]() {
  if (range_offset != 0) {
    size_t const offset = find_first_row_start(h_uncomp_data, h_uncomp_size);
    parse_csv_header(h_uncomp_data + offset, h_uncomp_size - header_offset);
    return offset;
  } else {
    return parse_csv_header(h_uncomp_data, h_uncomp_size);
  }
}();
```

OlivierNV replied:
The top one works for me

OlivierNV replied:
Done

OlivierNV commented on cpp/src/io/csv/reader_impl.cu:
```diff
@@ -346,194 +353,216 @@ table_with_metadata reader::impl::read(size_t range_offset,
+        const std::string quotechar(1, opts.quotechar);
+        const std::string dblquotechar(2, opts.quotechar);
+        std::unique_ptr<column> col =
+          make_column(column_types[i], num_records, out_buffers[i], stream, mr_);
```
Works for me

trevorsm7 commented on cpp/src/io/utilities/parsing_utils.cuh:
```diff
@@ -153,6 +153,11 @@ parse_numeric(const char* data, long start, long end, ParseOptions const& opts,
   }
 
+  // Skip over the "0x" prefix for hex notation
+  if (base == 16 && start + 2 <= end && data[start] == '0' && data[start + 1] == 'x') {
```
Maybe it would be a good to add an optional prefix string to ParseOptions, so you could also potentially support things like `#` in hex color strings (`#7400ff`), `0b` in binary strings, etc

OlivierNV replied:
Perhaps, but I'm definitely not doing this as part of this PR :) ~ really, the profile results show that the parsing code needs a complete overhaul, so that would be a good time to add functionality (right now, the primary goal is to bring the libcudf++ reader on par with the legacy reader)

trevorsm7 replied:
Also, does pandas automatically parse hex numbers when the prefix is present or does it require the base/hex to be explicitly specified? It seems like it would be easy to automatically override the base to 16 here if the first two chars are `0x`, and otherwise parse as base 10 or whatever, for example.

OlivierNV replied:
I would assume pandas infers the base automatically based on the prefix for integer columns, but cudf has explicit `"hex" / "hex32" / "hex64"` dtypes (probably something that should be revisited for improving pandas compatibility)

vuule replied:
I'm not sure is 0x is required for hex numbers. Users can specify a column to be hex, so maybe in that case the prefix is not needed.

vuule commented on cpp/src/io/csv/csv_gpu.h:
```diff
@@ -23,6 +23,112 @@ namespace cudf {
+/**
+ * @brief Select a row context given another input context (updating 62-bit row count)
+ **/
+inline __host__ __device__ uint64_t select_row_context(uint64_t sel_ctx, uint64_t packed_ctx)
```
I cannot figure out what this function does. Can you please explain in more detail in the comment above?

OlivierNV replied:
Added a more descriptive comment

vuule commented on cpp/src/io/csv/csv_gpu.h:
```diff
@@ -23,6 +23,112 @@ namespace cudf {
 namespace csv {
 namespace gpu {
+
+/**
```
I think it would help if we had aliases for the row context types (20-bit, 32-bit, packed, other(?)), so it's obvious what kind of context is passed into which function, and what kind is returned. 
This also provides a good spot to specify the content of these contexts.

OlivierNV replied:
There is really only the "normal" 32-bit version, eg: `row_count*4 + id` (which can be zero extended to 64-bit like any other number) or the 4x20-bit (packed) version. Would a `typedef uint64_t packed_rowcontext_t;` and `typedef uint32_t rowcontext_t` work ?

OlivierNV replied:
Went with yes: added `rowctx32_t`, `rowctx64_t` and `packed_rowctx_t` . It does make it easier to infer the intent without having to read the comments.

OlivierNV commented on cpp/src/io/csv/csv_gpu.h:
```diff
@@ -23,6 +23,112 @@ namespace cudf {
+ * @brief pack multiple row contexts together
+ *
+ * The 64-bit packed row context format represents the four possible output row context state
+ * from each of the four possible input row context state.
```
The fourth parsing context is `ROW_CTX_EOF` (this dead-end state is reached when reaching the last character)

OlivierNV commented on cpp/src/io/csv/csv_gpu.h:
```diff
@@ -23,6 +23,112 @@ namespace cudf {
+}
+
+/**
+ * @brief Unpack a row context  (select one of the 4 contexts in packed form)
```
It is valid to select the fourth context (not treated differently than the other 3 contexts)

trevorsm7 commented on cpp/src/io/csv/reader_impl.cu:
```diff
@@ -346,194 +351,215 @@ table_with_metadata reader::impl::read(size_t range_offset,
+  size_t header_end   = 0;
+  size_t data_start   = 0;
+  cudf::size_type row = 0;
+  int ctx             = ROW_CTX_NONE;
```
It might be helpful to call this a state or something to disambiguate it from the 20-bit context (row count + id/row_ctx) as used elsewhere.

OlivierNV replied:
Ok, went with "state". Oh wait, I just realized I renamed the wrong variable in csv_gpu.cpp. Gimme a sec.

OlivierNV replied:
Done. "state" definitely makes it clear that it's not quite the same thing as in the kernel.

trevorsm7 commented on cpp/src/io/csv/csv_gpu.h:
```diff
@@ -23,6 +23,120 @@ namespace cudf {
+ **/
+inline __host__ __device__ uint32_t make_row_context(uint32_t row_count, uint32_t out_ctx)
+{
+  return row_count * 4 + out_ctx;
```
```suggestion
  return row_count << 2 + out_ctx;
```

OlivierNV replied:
You don't want that (operator precedence).

trevorsm7 replied:
Ah, of course. How about `(row_count << 2) | out_ctx` then?

OlivierNV replied:
Sure. Either way is identical, so I don't have a strong preference for one way or the other (I'll generally just go for whatever results in the less amount of typing).

OlivierNV commented on cpp/src/io/csv/reader_impl.cu:
```diff
@@ -346,194 +351,215 @@ table_with_metadata reader::impl::read(size_t range_offset,
+        ctx = ROW_CTX_QUOTE;
+      }
+    } else if (ctx == ROW_CTX_QUOTE) {
+      if (ch == quotechar || (ch == esc_delimiter && prev_ch != escapechar)) { ctx = ROW_CTX_NONE; }
```
It's mostly functional here in the low-level kernel (at least for the purpose of escaping the delimiter), but not currently supported in the api. The function that launches the kernel has the correct code commented out, and I *think* I added a TODO comment somewhere (the header parsing also has disabled escapechar support)

OlivierNV commented on cpp/src/io/csv/reader_impl.cu:
```diff
@@ -346,194 +351,215 @@ table_with_metadata reader::impl::read(size_t range_offset,
+      if (prev_ch == opts.terminator && ch == commentchar) {
+        ctx = ROW_CTX_COMMENT;
+      } else if (ch == quotechar && (prev_ch == opts.terminator || prev_ch == opts.delimiter ||
+                                     prev_ch == quotechar)) {
```
Yeah. That's one illegal case where we would likely fail differently than pandas (doublequotes are only allowed in quoted strings or if quoting is disabled). Not planning to handle this particular case in this PR (the list of corner-cases is almost never-ending).
At some point soon, we'll need to deal with pandas compatibility for both the row-parsing stage and the data conversion stage (we should be able to match pandas in most cases ~ for the row parser, this potentially means looking at a larger window besides just ch_prev).

trevorsm7 replied:
Fair enough; I was just looking for ways to break the parser, but this would be illegal input as you say. Anyway, I suppose it would be easier to throw an exception when a quote is encountered and the previous char is _not_ a terminator, delimiter, or quote (which would fail at the first quote in `foo"",bar`). Assuming you want to fail when invalid input is encountered, rather than "garbage in garbage out", of course.

OlivierNV replied:
The correct behavior would be to match pandas (which might be hard in the case of column names). You often can't rule out invalid until you parse all the way to the end of the file (or give up after seeing a column name with a length greater than one million characters)

OlivierNV replied:
[For that particular case, I think the simplest solution would be to add a lookahead of the next character, so we don't change the quoting state when encountering double-quotes]

OlivierNV replied:
Will mark this as resolved for now: if we hit issues with ignoring quotes that do not start immediately after the delimiter, we could easily revert to the simple 2-state quoting scheme in the previous implementation by simply commenting out the delimiter-before-opening-quote check (the good thing about this new implementation is that it makes it possible to add more features to the parser while remaining fully data-parallel). I think it should be possible to have a data-parallel parser that fully matches pandas behavior (also hoping this is doable without adding more parser states ~ maybe a couple of flags/sub-states that would only impact the first state transition, or just brute-force look-ahead/look-behind for rare corner cases).

trevorsm7 commented on cpp/src/io/csv/csv_gpu.cu:
```diff
@@ -570,6 +571,329 @@ __global__ void convertCsvToGdf(const char *raw_csv,
+{
+  uint64_t tmp;
+
+#define CTX_MERGE(lanemask, tmask, base, level_scale)                       \
```
I think this could be a lambda function and still generate the same/similar inlined code. Not requesting a change, just a thinking out loud.

OlivierNV replied:
It'd be ok with me, but I'd want to see proof at godbolt that it does :) (I tend to not trust the compiler by default for this sort of things) ~ And also that it doesn't double the lines of code, otherwise I don't see the point.

trevorsm7 replied:
Right, just thinking out loud, not actually requesting a change. I know some people prefer to avoid macros when at all possible, so I mentioned it.

OlivierNV replied:
I know some people prefer to avoid lamdbas when at all possible [*insert "I know him, he is me" meme] :)

harrism replied:
Even a `constexpr inline` function would obviously be inlined -- the inputs are all literals.

OlivierNV replied:
Could also be a macro. It seems to me that this is the solution that results in the least amount of code. Less code = less bugs.

vuule replied:
What is the additional code that an inline function would have over the macro? I can only think of the parameter types, which is a positive IMO.

trevorsm7 commented on cpp/src/io/csv/csv_gpu.cu:
```diff
@@ -570,6 +571,329 @@ __global__ void convertCsvToGdf(const char *raw_csv,
+      c = cur[0];
+      if (c_prev == terminator) {
+        if (c == commentchar ||
+            (skipblanklines && (c == terminator || (c == '\r' && terminator == '\n')))) {
```
Question about `(c == '\r' && terminator == '\n')`. What if the character following the `\r` is not a `\n`? Wouldn't that cause `\r` to act the same as the comment escape char?

OlivierNV replied:
Yes, I noticed that too in the original implementation, so assuming it's ok for now, but probably will need to be revisited when addressing pandas compatibility (it's certainly convenient for now since it doesn't require looking up the next character, which will almost certainly be needed for matching pandas).

trevorsm7 replied:
I suppose this would be an unusual edge case. Would a lone non-terminator `\r` or other control character at the start of a line be considered invalid CSV? On the other hand, if this is unusual but valid CSV, then it seems like a bug.

OlivierNV replied:
I think the original implementation just assumes '\r' is always followed by '\n' (dos line endings), so it's almost certainly a bug unless pandas also does this.

OlivierNV replied:
The blank line removal is now in its own count_if/remove_if block (can use some suggestions there if we can efficiently share the is_blank lambda).

trevorsm7 commented on cpp/src/io/csv/csv_gpu.cu:
```diff
@@ -570,6 +571,329 @@ __global__ void convertCsvToGdf(const char *raw_csv,
+        }
+      } else if (c == quotechar) {
+        if (c_prev == delimiter || c_prev == quotechar) {
+          // Quoted string after delimiter, quoted string ending in delimiter, or double-quote
```
Have the pandas style double quotes already been replaced by single quotes at this point?

OlivierNV replied:
No. Basically when there is a double-quote inside a string, it's treated as two back-to-back strings here (leaves then re-enters quote state). If strings are not quoted, then there is no quotechar. Only allowing opening quotes when following a delimiter somewhat improves pandas compatibility for situations when quoting is incorrectly enabled (happens often since quoting is enabled by default -> in this case, we still break when there is a doublequote inside a non-quoted field, but it's presumably more rare)

trevorsm7 replied:
Oh, the double-quote issue is when there is a pair of quotes in the middle of a string, like `"foo""bar"` getting parsed as a single string? I thought it was that strings could have double quotes around them, like `""foobar""`.
Ok, and then `foo""bar` would be parsed incorrectly, but this seems to be invalid input anyway?

OlivierNV replied:
My understanding is that `"foo""bar"` gets parsed as the single-string string `foo"bar` when double-quotes are enabled. The data string `"foobar"` would be encoded in csv as `"""foobar"""` if quoting and doublequotes are enabled. I'm not entirely sure what happens if doublequotes are not enabled and string fields have quotes in them -> maybe the parser just looks for the next terminator ? (I suspect we'll need to have a lookahead loop to deal with such corner cases). `foo""bar`, `foo "bar" foo` and `11"W x 13"H x 10"D` would all be invalid if quoting is enabled, but I think the last two are more common (I think pandas automatically switches off quoting if the field doesn't start with a quote, so we may just need something like an additional quoted-field info bits if we want to avoid lookahead, at the cost of a slightly more complex merge op).

OlivierNV replied:
[Let's save these discussions for upcoming pandas compatibility work]

trevorsm7 commented on cpp/src/io/csv/csv_gpu.cu:
```diff
@@ -570,6 +571,329 @@ __global__ void convertCsvToGdf(const char *raw_csv,
+
+  // Convert the long-form {rowmap,outctx}[inctx] version into packed version
+  // {rowcount,ouctx}[inctx]
+  ctxb = pack_row_contexts(make_row_context(__popc(ctx_map.x), (ctx_map.w >> 0) & 3),
```
This could be an inlined utility function, `pack_short_row_context` or something

trevorsm7 replied:
```cpp
ctxb = pack_row_contexts(make_row_context(__popc(ctx_map.x), (ctx_map.w >> 0) & 3),
                           make_row_context(__popc(ctx_map.y), (ctx_map.w >> 2) & 3),
                           make_row_context(__popc(ctx_map.z), (ctx_map.w >> 4) & 3));
```

OlivierNV replied:
I just removed `ctxb` in the latest commit, and just called passed the result of `pack_row_context()` directly to `rowctx_merge_transform`, since it was used only for that purpose and only in one place.

OlivierNV replied:
Would still prefer to see a `pack_short_row_context` and `select_short_row_context` (or maybe `select_rowmap` ?

trevorsm7 replied:
I think it would be good to encapsulate all the low level bit twiddling stuff into short functions, yes. `select_rowmap` sounds good. I think the 32-bit contexts, packed contexts, short context, etc could be bundled up into classes without adding any overhead, but it could be done in a follow up PR (I wouldn't mind playing around with it myself)

Is this the only place where `rowctx_merge_transform` is called? I suppose you could pass `ctx_map` here instead and do the packing inside `rowctx_merge_transform`.

OlivierNV replied:
Yes, there is only one place where it is called, but the merge transform is a fairly generic operation whereas the short row context stuff is definitely specific to the calling kernel, so I'd rather keep them separate -> I'd rather go with the short util function in this case.

OlivierNV replied:
Went with `pack_rowmaps()` and `select_rowmap()` to avoid more "row context" confusion. Only concern is if the compiler would infer that ctx_map.w is still needed because the whole uint4 is being passed in to a function (+1 register), but not concerned enough to spend more time on this.

trevorsm7 commented on cpp/src/io/csv/csv_gpu.cu:
```diff
@@ -570,6 +571,329 @@ __global__ void convertCsvToGdf(const char *raw_csv,
+    rowctx32_t ctx             = rowctx_inverse_merge_transform(ctxtree, t);
+    uint64_t row               = (ctxtree[0] >> 2) + (ctx >> 2);
+    uint32_t rows_out_of_range = 0;
+    uint32_t rowmap =
```
This could be an inline function `select_short_row_context()` maybe?

trevorsm7 replied:
Something like
```cpp
uint32_t select_short_row_context_count(uint4 ctx_map, rowctx32_t ctx) {
  return ((ctx & 3) == ROW_CTX_NONE)
        ? ctx_map.x
        : ((ctx & 3) == ROW_CTX_QUOTE) ? ctx_map.y : ((ctx & 3) == ROW_CTX_COMMENT) ? ctx_map.z : 0;
}
```

OlivierNV replied:
Done.

OlivierNV commented on cpp/src/io/csv/csv_gpu.h:
```diff
@@ -23,6 +23,143 @@ namespace cudf {
+{
+  uint32_t ctxid = static_cast<uint32_t>(sel_ctx & 3);
+  rowctx32_t ctx = get_row_context(packed_ctx, ctxid);
+  return sel_ctx - ctxid + ctx;
```
Will experiment ~ was hoping compiler could use IADD3 (I almost added a comment `"// (sel_ctx & ~3) + ctx would also work" :)

OlivierNV replied:
Done

OlivierNV commented on cpp/src/io/csv/csv_gpu.cu:
```diff
@@ -570,6 +571,329 @@ __global__ void convertCsvToGdf(const char *raw_csv,
+
+  // Initial state is neutral context, zero rows
+  ctx_map.x = ctx_map.y = ctx_map.z = 0;
+  ctx_map.w                         = (0 << 0) | (1 << 2) | (2 << 4) | (3 << 6);  // i << (2*i)
```
Will do (likely needs clang-format pass)

OlivierNV replied:
Done

OlivierNV commented:
Crap. Missed something in the previous implementation: as weird as it is, skipped rows must be skipped *before* the header rows (causing python test failures). Another weird thing is that blank lines are not ignored by skiprows, so need to be removed in a separate pass. This also implies header parsing should happen after the row offset gathering phase.

OlivierNV commented:
Updated to drop blank rows only after applying skip_rows, and only then can header rows be dropped (it really needs to happen in a very specific order to match pandas).
That means I had to move the blank row removal to a separate count_if/remove_if pass (which is expected to not actually do anything in the vast majority of cases). Also removed the SW header parsing, since it is counter-intuitively located after skiprows and not always in the beginning of the file.

trevorsm7 commented on cpp/src/io/csv/csv_gpu.cu:
```diff
@@ -570,6 +577,376 @@ __global__ void convertCsvToGdf(const char *raw_csv,
+ * 1-bit count (0 or 1) per context in the lower 4 bits
+ * 2-bit output context id per input context in bits 8..15
+ **/
+inline __device__ uint32_t make_char_context(uint32_t id0,
```
Might be good to add `constexpr` here since this is expected to be evaluated at compile time

OlivierNV replied:
Done

OlivierNV commented:
@trevorsm7 FWIW, `(sel_ctx & ~3) + ctx` did save one instruction over `sel_ctx - ctxid + ctx`

OlivierNV commented:
> @trevorsm7 FWIW, `(sel_ctx & ~3) + ctx` did save one instruction over `sel_ctx - ctxid + ctx`

Arr, crap no: it saved one instruction in ptx, but it is the other way around in the real sm_70 binary, where `iadd3(a,b,-c)` indeed saves one instruction over the `add(a &~3, b)`

harrism commented on cpp/src/io/csv/csv_gpu.h:
```diff
@@ -16,13 +16,179 @@
+/**
+ * @brief pack multiple row contexts together
+ *
+ * The 64-bit packed row context format represents the four possible output row context state
```
```suggestion
 * The 64-bit packed row context format represents the four possible output row context states
```

harrism commented on cpp/src/io/csv/csv_gpu.h:
```diff
@@ -16,13 +16,179 @@
+ * @brief pack multiple row contexts together
+ *
+ * The 64-bit packed row context format represents the four possible output row context state
+ * from each of the four possible input row context state.
```
```suggestion
 * from each of the four possible input row context states.
```

harrism commented on cpp/src/io/csv/csv_gpu.cu:
```diff
@@ -488,15 +494,16 @@ struct decode_op {
-                                void **data,
-                                cudf::bitmask_type **valid)
+__global__ void __launch_bounds__(csvparse_block_dim)
+  convertCsvToGdf(const char *raw_csv,
```
```suggestion
  convert_csv_to_gdf(const char *raw_csv,
```

harrism commented on cpp/src/io/csv/csv_gpu.cu:
```diff
@@ -174,13 +179,14 @@ __device__ __inline__ bool is_floatingpoint(
-                                  const uint64_t *recStart,
-                                  column_parse::stats *d_columnData)
+__global__ void __launch_bounds__(csvparse_block_dim)
+  dataTypeDetection(const char *raw_csv,
```
While you are here, to match libcudf style...

```suggestion
  detect_data_types(const char *raw_csv,
```

harrism commented on cpp/src/io/csv/csv_gpu.cu:
```diff
@@ -570,6 +577,376 @@ __global__ void convertCsvToGdf(const char *raw_csv,
+                           get_row_context(second_ctx, id2));
+}
+
+/**
```
Low level functions should have doc blocks like this, but since they don't need to be in the public docs, don't start them with the doxygen indicator. (Suggest making this change to all the low-level functions in this PR -- we'll be working on documentation consistency across libcudf from here on.)

```suggestion
/*
```

harrism commented on cpp/src/io/csv/csv_gpu.cu:
```diff
@@ -580,10 +957,8 @@ cudaError_t __host__ DetectColumnTypes(const char *data,
-  CUDA_TRY(cudaOccupancyMaxPotentialBlockSize(&minGridSize, &blockSize, dataTypeDetection));
-  const int gridSize = (num_rows + blockSize - 1) / blockSize;
+  const int blockSize = csvparse_block_dim;
+  const int gridSize  = (num_rows + blockSize - 1) / blockSize;
```
```suggestion
  const cudf::experimental::detail::grid_1d grid{num_rows, block_size, 1};
```

OlivierNV replied:
I prefer the original code here

harrism replied:
I know that, and in most of my life I embrace diversity. But in libcudf I embrace consistency.

OlivierNV replied:
You mean you demand conformity ?

harrism commented on cpp/src/io/csv/csv_gpu.cu:
```diff
@@ -580,10 +957,8 @@ cudaError_t __host__ DetectColumnTypes(const char *data,
-  int minGridSize = 0;  // minimum block count required
-  CUDA_TRY(cudaOccupancyMaxPotentialBlockSize(&minGridSize, &blockSize, dataTypeDetection));
-  const int gridSize = (num_rows + blockSize - 1) / blockSize;
+  const int blockSize = csvparse_block_dim;
```
underscores, not camels.

```suggestion
  const int block_size = csvparse_block_dim;
```

harrism commented on cpp/src/io/csv/csv_gpu.cu:
```diff
@@ -603,17 +978,51 @@ cudaError_t __host__ DecodeRowColumnData(const char *data,
-  int minGridSize = 0;  // minimum block count required
-  CUDA_TRY(cudaOccupancyMaxPotentialBlockSize(&minGridSize, &blockSize, convertCsvToGdf));
-  const int gridSize = (num_rows + blockSize - 1) / blockSize;
+  const int blockSize = csvparse_block_dim;
```
```suggestion
  const int block_size = csvparse_block_dim;
```

harrism commented on cpp/src/io/csv/csv_gpu.cu:
```diff
@@ -603,17 +978,51 @@ cudaError_t __host__ DecodeRowColumnData(const char *data,
-  CUDA_TRY(cudaOccupancyMaxPotentialBlockSize(&minGridSize, &blockSize, convertCsvToGdf));
-  const int gridSize = (num_rows + blockSize - 1) / blockSize;
+  const int blockSize = csvparse_block_dim;
+  const int gridSize  = (num_rows + blockSize - 1) / blockSize;
```
```suggestion
  const cudf::experimental::detail::grid_1d grid{num_rows, block_size, 1};
```

harrism commented on cpp/src/io/csv/csv_gpu.cu:
```diff
@@ -570,6 +577,376 @@ __global__ void convertCsvToGdf(const char *raw_csv,
+  row_offsets.resize(new_end - row_offsets.begin());
+}
+
 cudaError_t __host__ DetectColumnTypes(const char *data,
```
```suggestion
cudaError_t __host__ detect_column_types(const char *data,
```

trevorsm7 commented:
Olivier, when did they implant an nvcc chip in your brain and how can I get one? Your intuition for translating C code to sm instructions is spooky.

OlivierNV commented on cpp/src/io/csv/csv_gpu.cu:
```diff
@@ -570,6 +577,376 @@ __global__ void convertCsvToGdf(const char *raw_csv,
+              (ROW_CTX_EOF << 6);  // No state transitions (pass-through context)
+  c_prev = (cur > start) ? cur[-1] : terminator;
+  // Loop through all 32 bytes and keep a bitmask of row starts for each possible input context
+  for (uint32_t pos = 0; pos < 32; pos++, cur++, c_prev = c) {
```
There are no scenarios where c could be uninitialized before it is used.

OlivierNV commented on cpp/src/io/csv/csv_gpu.h:
```diff
@@ -16,13 +16,179 @@
+/**
+ * @brief pack multiple row contexts together
+ *
+ * The 64-bit packed row context format represents the four possible output row context state
```
It's more like one is v2 of the same comment that was added along with the typedef addition, but the first sentence is indeed more appropriate when swapped around.

OlivierNV commented on cpp/src/io/csv/csv_gpu.h:
```diff
@@ -16,13 +16,179 @@
+ * states are included as parameters, and the EOF->EOF state transition is hardcoded)
+ *
+ **/
+inline __host__ __device__ packed_rowctx_t pack_row_contexts(rowctx32_t ctx0,
```
Regardless of what the sales brochure says, `constrexpr` is largely cosmetic (the problem of reducing constant expressions at compile time was solved decades ago). AFAICT, the primary use of constexpr is for compile-time asserts, though it may act as a sort of encouragement for the compiler for complex functions (such as loops/recursivity) where it may have given up attempting compile-time evaluation (clearly not the case here)
I can make it constexpr, though (why not, I suppose, it makes zero difference)

OlivierNV commented on cpp/src/io/csv/csv_gpu.cu:
```diff
@@ -570,6 +577,376 @@ __global__ void convertCsvToGdf(const char *raw_csv,
+  size_t block_pos =
+    (parse_pos - start_offset) + blockIdx.x * static_cast<size_t>(rowofs_block_bytes) + t * 32;
+  const char *cur = start + block_pos;
+  uint4 ctx_map;
```
That bug would get caught real quick, but okay.

OlivierNV commented:
Alright, let's get this csv show on the road.
